### PR TITLE
refactor and fix paxos batching

### PR DESCRIPTION
**Pull request by rjullman**

### CODE REVIEW

**Comment by rjullman** [comment]
The old code mimicked the batching in https://github.com/palantir/atlasdb/blob/develop/timestamp-impl/src/main/java/com/palantir/timestamp/RateLimitedTimestampService.java.  Unfortunately this did not not work (as written) for paxos since callers were allowed to join a batch at any time where as callers should only be able to join a batch before the check for leadership.  This is probably a little more of a rewrite than necessary, but I think this is in the right direction.  It is possible to keep the code more similar to before but that would require maintain a second map to keep track of queued callers.

This code is not tested, does not particular carefully consider the failure cases, and probably needs some more attention to thrashing.  I just wanted to get this up -- more revisions coming.

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    private void sendRequests() {
+        while (true) {
+            if (queue.isEmpty()) {
+                continue;
+            }
+
+            for (Map.Entry<LeadershipToken, Batch> e : queue.entrySet()) {
+                populateBatch(e.getValue(), e.getKey());
+            }
```
Comment (schlosna) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java]: Do we want to catch Throwable to ensure this method never terminates and kills the task/thread?

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+    public PaxosRequestBatcher(ExecutorService executor, Function<LeadershipToken, StillLeadingStatus> operation) {
+        this.queue = Maps.newConcurrentMap();
+        this.operation = operation;
+        executor.execute(() -> sendRequests());
+    }
+
+    private void sendRequests() {
+        while (true) {
+            if (queue.isEmpty()) {
```
Comment (schlosna) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java]: This is going to be a tight hot loop, do we need to sleep and backoff?

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        } finally {
+            batch.populationLatch.countDown();
+        }
+    }
+
+    private Batch getBatch(LeadershipToken key) {
+        Batch batch = queue.get(key);
+        while (batch == null) {
+            queue.putIfAbsent(key, new Batch());
+            batch = queue.get(key);
```
Comment (schlosna) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java]: `batch = queue.computeIfAbsent(key, () -> new Batch());`

File: leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        }
+        return batch;
+    }
+
+    private static class Batch {
+        public final CountDownLatch populationLatch;
+        // should be written by 1 thread (the batching thread) and read by the rest
+        public volatile StillLeadingStatus ret;
+        public volatile Throwable error = null;
```
Comment (schlosna) [inline leader-election-impl/src/main/java/com/palantir/leader/PaxosRequestBatcher.java]: let's encapsulate these

**Comment by rjullman** [comment]
@schlosna -- most of your comments here are on the old code (stuff that was just moved).  I'm putting up a new lighter PR that will cleanup the old code (e.g. encapsulation) and probably be easier to review.

**Comment by rjullman** [comment]
https://github.com/palantir/atlasdb/pull/273 is a simpler alternative to this

**Pull request closed without merging**